File: languages/php/Bloock/Entity/Identity/CredentialOffer.php
<?php

namespace Bloock\Entity\Identity;

use Bloock\Bridge\Bridge;
use Bloock\Config\Config;
use Bloock\CredentialOfferFromJsonRequest;
use Bloock\CredentialOfferToJsonRequest;
use Exception;

class CredentialOffer
{
    private string $thid;
    private CredentialOfferBody $body;
    private string $from;
    private string $to;

    /**
     * @param string $thid
     * @param CredentialOfferBody $body
     * @param string $from
     * @param string $to
     */
    public function __construct(string $thid, CredentialOfferBody $body, string $from, string $to)
    {
        $this->thid = $thid;
        $this->body = $body;
        $this->from = $from;
        $this->to = $to;
    }

    /**
     * @throws Exception
     */
    static public function fromJSON(string $json): CredentialOffer
    {
        $bridge = new Bridge();

        $req = new CredentialOfferFromJsonRequest();
        $req->setConfigData(Config::newConfigDataDefault());
        $req->setJson($json);

        $res = $bridge->identity->CredentialOfferFromJson($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return CredentialOffer::fromProto($res->getCredentialOffer());
    }

    public static function fromProto(\Bloock\CredentialOffer $res): CredentialOffer
    {
        return new CredentialOffer($res->getThid(), CredentialOfferBody::fromProto($res->getBody()), $res->getFrom(), $res->getTo());
    }

    /**
     * @throws Exception
     */
    public function toJSON(): string
    {
        $bridge = new Bridge();

        $req = new CredentialOfferToJsonRequest();
        $req->setConfigData(Config::newConfigDataDefault());
        $req->setCredentialOffer($this->toProto());

        $res = $bridge->identity->CredentialOfferToJson($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return $res->getJson();
    }

    public function toProto(): \Bloock\CredentialOffer
    {
        $p = new \Bloock\CredentialOffer();
        $p->setThid($this->thid);
        $p->setBody($this->body->toProto());
        $p->setFrom($this->from);
        $p->setTo($this->to);
        return $p;
    }
}

File: languages/python/bloock/entity/identity/credential_offer.py
from __future__ import annotations

import bloock._bridge.proto.identity_entities_pb2 as proto
from bloock._bridge import BloockBridge
from bloock._bridge.proto.identity_pb2 import CredentialOfferFromJsonRequest, CredentialOfferToJsonRequest
from bloock._bridge.proto.shared_pb2 import Error
from bloock._config.config import Config
from bloock.entity.identity.credential_offer_body import CredentialOfferBody


class CredentialOffer:
    def __init__(self, thid: str, body: CredentialOfferBody, _from: str, _to: str) -> None:
        self.thid = thid
        self.body = body
        self._from = _from
        self._to = _to

    @staticmethod
    def from_json(json: str) -> CredentialOffer:
        bridge = BloockBridge()

        req = CredentialOfferFromJsonRequest(
            config_data=Config.default(),
            json=json
        )

        res = bridge.identity().CredentialOfferFromJson(req)
        if res.error != Error():
            raise Exception(res.error.message)

        return CredentialOffer.from_proto(res.credential_offer)

    def to_json(self) -> str:
        bridge = BloockBridge()

        req = CredentialOfferToJsonRequest(
            config_data=Config.default(),
            credential_offer=self.to_proto()
        )

        res = bridge.identity().CredentialOfferToJson(req)
        if res.error != Error():
            raise Exception(res.error.message)

        return res.json

    @staticmethod
    def from_proto(c: proto.CredentialOffer) -> CredentialOffer:
        return CredentialOffer(
            thid=c.thid,
            body=CredentialOfferBody.from_proto(c.body),
            _to=c._to,
            _from=c._from
        )

    def to_proto(self) -> proto.CredentialOffer:
        return proto.CredentialOffer(
            thid=self.thid,
            body=self.body.to_proto(),
            _to=self._to,
            _from=self._from
        )
